source/MaterialXCore/Interface.cpp

```diff
@@ -66,11 +162,22 @@ bool PortElement::validate(string* message) const
                 validateRequire(output != nullptr, res, message, "Invalid output in port connection");
                 if (output)
                 {
-                    validateRequire(getType() == output->getType(), res, message, "Mismatched output type in port connection");
+                    if (hasChannels())
+                    {
+                        validateRequire(supportsSwizzling(output->getType(), getType(), getChannels()), res, message, "Invalid channels attribute");
+                    }
+                    else
+                    {
+                        validateRequire(getType() == output->getType(), res, message, "Mismatched output type in port connection");
+                    }
                 }
             }
         }
-        else
+        else if (hasChannels())
+        {
+            validateRequire(supportsSwizzling(getConnectedNode()->getType(), getType(), getChannels()), res, message, "Invalid channels attribute");
```

Comment:
Note really part of this review, but just a note that getConnectedNode() is being called many times in this validate method. Since it's not a cheap operation we could do it once at the top and save the result.

Reply:
Did it in some other places so will clean up here.